### DISCUSSION TITLE
blog/series/auth/keycloak/[Auth]%20-%20keycloak_nextauth_spring_01

### DISCUSSION THREAD

**Discussion by giscus** [swj9707/swj_techblog#7] [Reply]
# blog/series/auth/keycloak/[Auth]%20-%20keycloak_nextauth_spring_01

Keycloak, NextAuth, SpringBoot 연동 가이드, Keycloak Setup 방법

https://swj-techblog.vercel.app/blog/series/auth/keycloak/[Auth]%20-%20keycloak_nextauth_spring_01

**Comment by crinity-khg** [comment]
안녕하세요.
혹시 next-auth 를 사용할 때 google 같은 provider 말고 포스트에 작성하신것처럼
직접 구성한 jwt 인증 서버를 사용 할 때 
로그인 이후에 jwt 를 서버에서 body 나 header 에 return 받는게 아니라
next-auth 에 jwt 콜백에서 서버로 요청을 보내서 jwt 를 반환받아 오는건가요??
혹시 가능하시면 next-auth 로 연계되는 시나리오를 알 수 있을까요..?

**Reply by oxspan** [reply]
안녕하세요 오랫동안 방치되었던 글에 관심을 가져주셔서 감사합니다. 
우선 죄송합니다. 너무 오래 이 주제로 글을 쓰지 않아서 궁금 해 하시는 영역에 대한 해소가 제대로 되지 않으셨군요 ㅠ

우선 직접 구성한 JWT 인증서버라기보단 Keycloak 오픈소스를 Google OAuth 와 같이 SSO 서버로 배치 한 상황입니다. 
next-auth 를 써보셔서 잘 아시겠지만 커스텀 Provider를 만들 수 있어요. 말씀 하셨던 것 처럼 jwtCallback, authorize, session 등의 콜백 함수들을 커스텀 하시면 됩니다. keycloak Provider 에서 authorize 가 호출될 때 keycloak 에 JWT 토큰 생성 요청을 보낸 후 리턴되는 값을 토대로 세션 정보가갱신되게 됩니다.  next-auth.d.ts 를 통해 세션에 들어오는 User, Session 인터페이스를 커스텀 할 수 있어요. 그 후 Spring Boot  등의 API 서버에서는 세션에 저장 된 토큰 정보를 바탕으로 통신을 하게 되는데 Spring Boot OAuth Resource Server 디펜던시를 활용하면 Keycloak 에서 생성 된 토큰을 검증하는 기능을 추가할 수 있어요. 

```typescript
import { keycloakSignIn, reissueToken } from "@/service/auth/AuthService";
import { TokenRes } from "@/types/auth/types";
import type { NextAuthOptions, Session, SessionUser, User } from "next-auth";
import NextAuth from "next-auth";
import { JWT } from "next-auth/jwt";
import CredentialsProvider from "next-auth/providers/credentials";

export const jwtCallback = async ({
  token,
  user,
}: {
  token: JWT;
  user: User;
}) => {
  if (token?.expiresIn) {
    if (Math.floor(Date.now() / 1000) < token.expiresIn!!) {
      return { ...token, ...user };
    } else if (
      token.refreshToken &&
      Math.floor(Date.now() / 1000) < token.refreshExpiresIn!!
    ) {
      const newTokenData: TokenRes = await reissueToken(token.refreshToken);
      token.accessToken = newTokenData.accessToken;
      token.refreshToken = newTokenData.refreshToken;
      token.expiresIn = newTokenData.expiresIn;
      token.refreshExpiresIn = newTokenData.refreshExpiresIn;
      return { ...token, ...newTokenData };
    }
  }

  token.accessToken = user.tokenRes.accessToken;
  token.refreshToken = user.tokenRes.refreshToken;
  token.expiresIn = user.tokenRes.expiresIn;
  token.refreshExpiresIn = user.tokenRes.refreshExpiresIn;
  const sessionUser: SessionUser = {
    id: user.id,
    roles: user.roles,
  };
  return { ...token, ...sessionUser };
};

export const session = ({
  session,
  token,
}: {
  session: Session;
  token: JWT;
}): Promise<Session> => {
  if (
    Math.floor(Date.now() / 1000) > token?.expiresIn! &&
    token?.refreshTokenExpires &&
    Math.floor(Date.now() / 1000) > token?.refreshExpiresIn!
  ) {
    return Promise.reject({
      error: new Error("토큰이 만료되었습니다. 재 로그인 해 주세요."),
    });
  }

  session.tokenInfo = token;
  return Promise.resolve(session);
};

export const authOptions: NextAuthOptions = {
  providers: [
    CredentialsProvider({
      id: "keycloak",
      name: "keycloak",
      credentials: {
        username: {
          label: "Username",
          type: "email",
          placeholder: "유저 이메일",
        },
        password: { label: "Password", type: "password" },
      },
      async authorize(credentials) {
        return await keycloakSignIn(
          credentials?.username as string,
          credentials?.password as string
        )
          .then((res) => {
            return res;
          })
          .catch(() => Promise.reject());
      },
    }),
  ],
  pages: {
    signIn: "/login",
  },
  callbacks: {
    jwt: jwtCallback,
    session: session,
  },
};
export default NextAuth(authOptions);
``` 
1년 전 코드고 그닥 좋은 코드라고 볼 수는 없지만, 도움이 되실거에요. 만약 Google OAuth Provider를 쓰신다면 공식 문서대로 따라가면 되겠지만 지금 같은 케이스에서는 Provider 내의 콜백들을 커스텀 해 주셔야 합니다. 

답변이 되셨을까요? 또 질문 있으시면 답글 남겨주시고 빠른 시일 내에 다시 글을 써 보겠습니다...이 주제로 글을 다시 써야지 생각은 항상 하는데 쉽지가 않군요 ㅠ 다 핑계겠죠

**Reply by crinity-khg** [reply]
전체적인 흐름이

사용자 생성, 로그인, 인증 전부 => keycloack 에서 담당하게 되는거고
백엔드 Spring 서버는 keycloack 에서 생성한 토큰만 유효한지 검증해서 요청 진행하게 하는거고

next-auth 에서는 로그인하면 keycloack 에서 생성한 토큰을 세션에 담아서 사용하는거고
세션 토큰에 담긴 jwt 에서 expiresIn 확인해서 만료 된 순간
작성해주신 코드에 콜백에서 token.refreshToken 으로 담아놓은 리프레쉬토큰을 꺼내서 keycloak 에 재발급요청을 보내는 흐름인게 맞을까요..??

next-auth 도 많이 사용해본게 아니라서 헷갈리네요 죄송합니다 ㅠ

**Reply by oxspan** [reply]
네 맞아요. 한번 만들어 보시면 금방 이해하실거에요. 프론트엔드 애플리케이션 서버, 인증서버, REST api 애플리케이션 서버 세가지가 통신하는 구조에요.
